foyer-memory/src/cache.rs

```diff
@@ -0,0 +1,601 @@
+        if let Some(old) = self.indexer.insert(ptr) {
+            debug_assert!(old.as_ref().base().is_in_eviction());
+            self.eviction.remove(old);
+            debug_assert!(!old.as_ref().base().is_in_eviction());
+            // Because the `old` handle is removed from the indexer, it will not be reinserted again.
+            if let Some(entry) = self.try_release_handle(old) {
+                last_reference_entries.push(entry);
+            }
+        }
+        self.eviction.push(ptr);
+
+        debug_assert!(ptr.as_ref().base().is_in_indexer());
+        debug_assert!(ptr.as_ref().base().is_in_indexer());
+
+        self.usage.fetch_add(charge, Ordering::Relaxed);
+        ptr.as_mut().base_mut().inc_refs();
+
+        ptr
+    }
+
+    unsafe fn get(&mut self, hash: u64, key: &K) -> Option<NonNull<H>> {
+        let mut ptr = self.indexer.get(hash, key)?;
+        let base = ptr.as_mut().base_mut();
+        debug_assert!(base.is_in_indexer());
+
+        base.inc_refs();
+        self.eviction.access(ptr);
+
+        Some(ptr)
+    }
+
+    /// Remove a key from the cache.
+    ///
+    /// Return `Some(..)` if the handle is released, or `None` if the handle is still in use.
+    unsafe fn remove(&mut self, hash: u64, key: &K) -> Option<(K, V)> {
+        let ptr = self.indexer.remove(hash, key)?;
+        self.try_release_handle(ptr)
+    }
+
+    /// Clear all cache entries.
+    unsafe fn clear(&mut self, last_reference_entries: &mut Vec<(K, V)>) {
+        // TODO(MrCroxx): Avoid collecting here?
+        let ptrs = self.indexer.drain().collect_vec();
+        let eptrs = self.eviction.clear();
+
+        // Assert that the handles in the indexer covers the handles in the eviction container.
+        if cfg!(debug_assertions) {
+            use std::{collections::HashSet as StdHashSet, hash::RandomState as StdRandomState};
+            let ptrs: StdHashSet<_, StdRandomState> = StdHashSet::from_iter(ptrs.iter().copied());
+            let eptrs: StdHashSet<_, StdRandomState> = StdHashSet::from_iter(eptrs.iter().copied());
+            debug_assert!((&eptrs - &ptrs).is_empty());
+        }
+
+        // The handles in the indexer covers the handles in the eviction container.
+        // So only the handles drained from the indexer need to be released.
+        for ptr in ptrs {
+            if let Some(entry) = self.try_release_handle(ptr) {
+                last_reference_entries.push(entry);
+            }
+        }
+    }
+
+    unsafe fn evict(&mut self, charge: usize, last_reference_entries: &mut Vec<(K, V)>) {
+        while self.usage.load(Ordering::Relaxed) + charge > self.capacity
+            && let Some(evicted) = self.eviction.pop()
+        {
+            let base = evicted.as_ref().base();
+            debug_assert!(base.is_in_indexer());
+            debug_assert!(!base.is_in_eviction());
+            if let Some(entry) = self.try_release_handle(evicted) {
```

Comment:
Discussed offline: `reinsert` in `try_release_handle` may be triggered, which will cause surprise, because `self.usage.load(Ordering::Relaxed) <= self.capacity` may not be held here.